dali/pipeline/operator/sequence_op.h

```diff
@@ -0,0 +1,471 @@
+    if (!expand_) {
+      is_inferred = Operator<Backend>::Setup(output_desc, ws);
+    } else {
+      SetupSequenceOperator(ws);
+      SetupExpandedWorkspace(expanded_, ws);
+      ExpandInputs(ws);
+      ExpandArguments(ws);
+      is_inferred = Operator<Backend>::Setup(output_desc, expanded_);
+    }
+    return ProcessOutputDesc(output_desc, ws, is_inferred);
+  }
+
+  void Run(workspace_t<Backend> &ws) override {
+    if (!expand_) {
+      Operator<Backend>::Run(ws);
+    } else {
+      ExpandOutputs(ws);
+      Operator<Backend>::Run(expanded_);
+      PostprocessOutputs(ws);
+      Clear();
+    }
+  }
+
+  bool IsExpanded() const {
+    return expand_;
+  }
+
+  const SampleFrameCtx &GetSampleFrameCtx() {
+    return sample_ctx_;
+  }
+
+  virtual bool ShouldExpandChannels() const {
+    return true;
+  }
+
+ protected:
+  virtual bool ShouldExpand(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      const auto &layout = GetInputLayout(ws, input_idx);
+      auto layout_desc =
+          ShouldExpandChannels() ? LayoutDesc::FrameAndChannel(layout) : LayoutDesc::Frame(layout);
+      if (layout_desc.NumDims() > 0) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  virtual void SetupSequenceOperator(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    input_expand_desc_.reserve(num_inputs);
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      const auto &input_shape = ws.GetInputShape(input_idx);
+      const auto &layout = GetInputLayout(ws, input_idx);
+      DALI_ENFORCE(layout.size() == 0 || layout.size() == input_shape.sample_dim(),
+                   make_string("Layout of input ", input_idx, " has size ", layout.size(),
+                               " which does not match the dimensionality of the input (got input "
+                               "with dimensionality ",
+                               input_shape.sample_dim(), ")."));
+      auto layout_desc =
+          ShouldExpandChannels() ? LayoutDesc::FrameAndChannel(layout) : LayoutDesc::Frame(layout);
+      input_expand_desc_.emplace_back(input_shape, layout_desc);
+    }
+  }
+
+  virtual void ExpandInputs(const workspace_t<Backend> &ws) {
+    for (int input_idx = 0; input_idx < ws.NumInput(); input_idx++) {
+      const auto &expand_desc = GetInputExpandDesc(input_idx);
+      auto num_expand_dims = expand_desc.NumDims();
+      if (ws.template InputIsType<GPUBackend>(input_idx)) {
+        auto expanded_handle = ExpandInputHelper<GPUBackend>(ws, input_idx, num_expand_dims);
+        AddInputHelper<GPUBackend>(expanded_handle, {ExpandDescFrameInfoFn{&expand_desc}});
+      } else {
+        auto expanded_handle = ExpandInputHelper<CPUBackend>(ws, input_idx, num_expand_dims);
+        AddInputHelper<CPUBackend>(expanded_handle, {ExpandDescFrameInfoFn{&expand_desc}});
+      }
+    }
+  }
```

Comment:
1. Do you check that the number of frames matches across the expanded inputs?
2. What happens when one input is expanded and another one isn't? Is there some broadcasting or is it an error?

Reply:
1. I don't validate them. Although, I believe, it is validated by `Operator<Backend>::Setup` call (EnforceUniformInputBatchSize). It seemed fine that the equal number of samples for every input is not enforced by this operator - it doesn't have to rely on that.
2. I deliberately left it unhandled - simply one would be expanded and the other one not -  but thinking again, maybe erroring out by default at the ShouldExpand level would be a better idea - after all it depends on the user input.